File: app/src/main/java/com/tencent/mobilenetssdncnn/MainActivity.java
package com.tencent.mobilenetssdncnn;

import android.Manifest;
import android.app.Activity;
import android.content.Context;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.graphics.SurfaceTexture;
import android.hardware.camera2.CameraAccessException;
import android.hardware.camera2.CameraCaptureSession;
import android.hardware.camera2.CameraCharacteristics;
import android.hardware.camera2.CameraDevice;
import android.hardware.camera2.CameraManager;
import android.hardware.camera2.CaptureRequest;
import android.hardware.camera2.params.StreamConfigurationMap;
import android.os.Bundle;
import android.os.Handler;
import android.os.HandlerThread;
import android.util.Log;
import android.util.Size;
import android.view.Surface;
import android.view.TextureView;
import android.view.View;
import android.view.WindowManager;
import android.widget.Button;
import android.widget.Toast;

import java.util.Collections;
import java.util.concurrent.Semaphore;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;

public class MainActivity extends Activity {
    private static final String TAG = "MainActivity";
    private static final int PERMISSION_REQUEST_CAMERA = 101;

    private TextureView cameraTextureView;
    private OverlayView overlayView;
    private Button btnToggleGpu;
    private Button btnSwitchCamera;
    private Button btnToggleInference;

    private final MobilenetSSDNcnn mobilenetssdncnn = new MobilenetSSDNcnn();

    private CameraDevice cameraDevice;
    private CameraCaptureSession captureSession;
    private CaptureRequest.Builder previewRequestBuilder;

    private HandlerThread backgroundThread;
    private Handler backgroundHandler;

    private HandlerThread inferenceThread;
    private Handler inferenceHandler;

    private String currentCameraId;
    private int facing = CameraCharacteristics.LENS_FACING_BACK;
    private boolean useGpu = false;
    private boolean inferenceEnabled = true;

    private final Semaphore cameraOpenCloseLock = new Semaphore(1);
    private final AtomicBoolean isInferencing = new AtomicBoolean(false);
    private Bitmap reusableBitmap = null;

    private float smoothedLatency = 0f;
    private float smoothedFps = 0f;
    private long lastFrameTimestamp = 0;

    private final TextureView.SurfaceTextureListener surfaceTextureListener = new TextureView.SurfaceTextureListener() {
        @Override
        public void onSurfaceTextureAvailable(SurfaceTexture surface, int width, int height) {
            openCamera();
        }

        @Override
        public void onSurfaceTextureSizeChanged(SurfaceTexture surface, int width, int height) {}

        @Override
        public boolean onSurfaceTextureDestroyed(SurfaceTexture surface) {
            return true;
        }

        @Override
        public void onSurfaceTextureUpdated(SurfaceTexture surface) {
            if (!inferenceEnabled || isInferencing.get()) {
                return;
            }

            if (isInferencing.compareAndSet(false, true)) {
                if (inferenceHandler != null) {
                    inferenceHandler.post(new Runnable() {
                        @Override
                        public void run() {
                            processCurrentFrame();
                        }
                    });
                } else {
                    isInferencing.set(false);
                }
            }
        }
    };

    private final CameraDevice.StateCallback stateCallback = new CameraDevice.StateCallback() {
        @Override
        public void onOpened(CameraDevice camera) {
            cameraOpenCloseLock.release();
            cameraDevice = camera;
            createCameraPreviewSession();
        }

        @Override
        public void onDisconnected(CameraDevice camera) {
            cameraOpenCloseLock.release();
            camera.close();
            cameraDevice = null;
        }

        @Override
        public void onError(CameraDevice camera, int error) {
            cameraOpenCloseLock.release();
            camera.close();
            cameraDevice = null;
            Log.e(TAG, "CameraDevice error: " + error);
        }
    };

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
        setContentView(R.layout.main);

        // Initialize ncnn MobileNet-SSD
        boolean retInit = mobilenetssdncnn.Init(getAssets());
        if (!retInit) {
            Log.e(TAG, "mobilenetssdncnn Init failed");
            Toast.makeText(this, "ncnn model init failed", Toast.LENGTH_LONG).show();
        } else {
            Log.i(TAG, "mobilenetssdncnn Init successful");
        }

        cameraTextureView = findViewById(R.id.cameraTextureView);
        overlayView = findViewById(R.id.overlayView);
        btnToggleGpu = findViewById(R.id.btnToggleGpu);
        btnSwitchCamera = findViewById(R.id.btnSwitchCamera);
        btnToggleInference = findViewById(R.id.btnToggleInference);

        btnToggleGpu.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                useGpu = !useGpu;
                btnToggleGpu.setText(useGpu ? "Mode: GPU" : "Mode: CPU");
                btnToggleGpu.setBackgroundColor(useGpu ? 0xFF9C27B0 : 0xFF3366CC);
            }
        });

        btnSwitchCamera.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                facing = (facing == CameraCharacteristics.LENS_FACING_BACK) ?
                        CameraCharacteristics.LENS_FACING_FRONT : CameraCharacteristics.LENS_FACING_BACK;
                btnSwitchCamera.setText(facing == CameraCharacteristics.LENS_FACING_BACK ? "Back Cam" : "Front Cam");
                if (backgroundHandler != null) {
                    backgroundHandler.post(new Runnable() {
                        @Override
                        public void run() {
                            closeCamera();
                            openCamera();
                        }
                    });
                }
            }
        });

        btnToggleInference.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                inferenceEnabled = !inferenceEnabled;
                btnToggleInference.setText(inferenceEnabled ? "Running" : "Paused");
                btnToggleInference.setBackgroundColor(inferenceEnabled ? 0xFF2E7D32 : 0xFFB71C1C);
                if (!inferenceEnabled) {
                    overlayView.setResults(null, 300, 300, 0f, 0f, useGpu,
                            facing == CameraCharacteristics.LENS_FACING_FRONT, "Paused");
                }
            }
        });
    }

    @Override
    protected void onResume() {
        super.onResume();
        startBackgroundThreads();

        if (checkSelfPermission(Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
            requestPermissions(new String[]{Manifest.permission.CAMERA}, PERMISSION_REQUEST_CAMERA);
        } else {
            if (cameraTextureView.isAvailable()) {
                openCamera();
            } else {
                cameraTextureView.setSurfaceTextureListener(surfaceTextureListener);
            }
        }
    }

    @Override
    protected void onPause() {
        closeCamera();
        stopBackgroundThreads();
        super.onPause();
    }

    private void startBackgroundThreads() {
        backgroundThread = new HandlerThread("CameraBackground");
        backgroundThread.start();
        backgroundHandler = new Handler(backgroundThread.getLooper());

        inferenceThread = new HandlerThread("InferenceWorker");
        inferenceThread.start();
        inferenceHandler = new Handler(inferenceThread.getLooper());
    }

    private void stopBackgroundThreads() {
        if (backgroundThread != null) {
            backgroundThread.quitSafely();
            try {
                backgroundThread.join();
                backgroundThread = null;
                backgroundHandler = null;
            } catch (InterruptedException e) {
                Log.e(TAG, "stopBackgroundThreads interrupted", e);
            }
        }
        if (inferenceThread != null) {
            inferenceThread.quitSafely();
            try {
                inferenceThread.join();
                inferenceThread = null;
                inferenceHandler = null;
            } catch (InterruptedException e) {
                Log.e(TAG, "stopInferenceThreads interrupted", e);
            }
        }
    }

    private void openCamera() {
        if (checkSelfPermission(Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
            return;
        }

        CameraManager manager = (CameraManager) getSystemService(Context.CAMERA_SERVICE);
        try {
            if (!cameraOpenCloseLock.tryAcquire(2500, TimeUnit.MILLISECONDS)) {
                throw new RuntimeException("Time out waiting to lock camera opening.");
            }

            currentCameraId = null;
            for (String id : manager.getCameraIdList()) {
                CameraCharacteristics characteristics = manager.getCameraCharacteristics(id);
                Integer lensFacing = characteristics.get(CameraCharacteristics.LENS_FACING);
                if (lensFacing != null && lensFacing == facing) {
                    currentCameraId = id;
                    break;
                }
            }

            if (currentCameraId == null && manager.getCameraIdList().length > 0) {
                currentCameraId = manager.getCameraIdList()[0];
            }

            manager.openCamera(currentCameraId, stateCallback, backgroundHandler);
        } catch (CameraAccessException e) {
            Log.e(TAG, "Cannot access camera", e);
            cameraOpenCloseLock.release();
        } catch (InterruptedException e) {
            throw new RuntimeException("Interrupted while trying to lock camera opening.", e);
        }
    }

    private void closeCamera() {
        try {
            cameraOpenCloseLock.acquire();
            if (captureSession != null) {
                captureSession.close();
                captureSession = null;
            }
            if (cameraDevice != null) {
                cameraDevice.close();
                cameraDevice = null;
            }
        } catch (InterruptedException e) {
            throw new RuntimeException("Interrupted while trying to lock camera closing.", e);
        } finally {
            cameraOpenCloseLock.release();
        }
    }

    private void createCameraPreviewSession() {
        try {
            SurfaceTexture texture = cameraTextureView.getSurfaceTexture();
            if (texture == null) return;

            CameraManager manager = (CameraManager) getSystemService(Context.CAMERA_SERVICE);
            CameraCharacteristics characteristics = manager.getCameraCharacteristics(currentCameraId);
            StreamConfigurationMap map = characteristics.get(CameraCharacteristics.SCALER_STREAM_CONFIGURATION_MAP);
            if (map != null) {
                Size[] sizes = map.getOutputSizes(SurfaceTexture.class);
                if (sizes != null && sizes.length > 0) {
                    Size optimalSize = sizes[0];
                    for (Size sz : sizes) {
                        if (sz.getWidth() <= 1280 && sz.getHeight() <= 720) {
                            optimalSize = sz;
                            break;
                        }
                    }
                    texture.setDefaultBufferSize(optimalSize.getWidth(), optimalSize.getHeight());
                }
            }

            Surface surface = new Surface(texture);
            previewRequestBuilder = cameraDevice.createCaptureRequest(CameraDevice.TEMPLATE_PREVIEW);
            previewRequestBuilder.addTarget(surface);
            previewRequestBuilder.set(CaptureRequest.CONTROL_MODE, CaptureRequest.CONTROL_MODE_AUTO);
            previewRequestBuilder.set(CaptureRequest.CONTROL_AF_MODE, CaptureRequest.CONTROL_AF_MODE_CONTINUOUS_PICTURE);
            previewRequestBuilder.set(CaptureRequest.CONTROL_AE_MODE, CaptureRequest.CONTROL_AE_MODE_ON);

            cameraDevice.createCaptureSession(Collections.singletonList(surface), new CameraCaptureSession.StateCallback() {
                @Override
                public void onConfigured(CameraCaptureSession session) {
                    if (cameraDevice == null) return;
                    captureSession = session;
                    try {
                        captureSession.setRepeatingRequest(previewRequestBuilder.build(), null, backgroundHandler);
                    } catch (CameraAccessException e) {
                        Log.e(TAG, "Failed to start camera preview", e);
                    }
                }

                @Override
                public void onConfigureFailed(CameraCaptureSession session) {
                    Log.e(TAG, "Camera configuration failed");
                }
            }, backgroundHandler);
        } catch (CameraAccessException e) {
            Log.e(TAG, "createCameraPreviewSession error", e);
        }
    }

    private void processCurrentFrame() {
        try {
            if (reusableBitmap == null || reusableBitmap.getWidth() != 300 || reusableBitmap.getHeight() != 300) {
                reusableBitmap = Bitmap.createBitmap(300, 300, Bitmap.Config.ARGB_8888);
            }

            Bitmap frame = cameraTextureView.getBitmap(reusableBitmap);
            if (frame == null) {
                isInferencing.set(false);
                return;
            }

            long startTime = System.currentTimeMillis();
            MobilenetSSDNcnn.Obj[] objects = mobilenetssdncnn.Detect(frame, useGpu);
            long endTime = System.currentTimeMillis();

            long latency = Math.max(1, endTime - startTime);
            if (smoothedLatency == 0) {
                smoothedLatency = latency;
            } else {
                smoothedLatency = 0.8f * smoothedLatency + 0.2f * latency;
            }

            if (lastFrameTimestamp > 0) {
                long frameInterval = Math.max(1, endTime - lastFrameTimestamp);
                float instantaneousFps = 1000f / frameInterval;
                if (smoothedFps == 0) {
                    smoothedFps = instantaneousFps;
                } else {
                    smoothedFps = 0.85f * smoothedFps + 0.15f * instantaneousFps;
                }
            }
            lastFrameTimestamp = endTime;

            final MobilenetSSDNcnn.Obj[] finalObjects = objects;
            final float curLatency = smoothedLatency;
            final float curFps = smoothedFps;
            final boolean curGpu = useGpu;
            final boolean isFront = (facing == CameraCharacteristics.LENS_FACING_FRONT);

            runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    overlayView.setResults(finalObjects, 300, 300, curLatency, curFps, curGpu, isFront, "Active");
                }
            });
        } catch (Exception e) {
            Log.e(TAG, "Error in processCurrentFrame", e);
        } finally {
            isInferencing.set(false);
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
        if (requestCode == PERMISSION_REQUEST_CAMERA) {
            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                if (cameraTextureView.isAvailable()) {
                    openCamera();
                } else {
                    cameraTextureView.setSurfaceTextureListener(surfaceTextureListener);
                }
            } else {
                Toast.makeText(this, "Camera permission is required for live detection", Toast.LENGTH_LONG).show();
            }
        }
    }
}

File: app/src/main/java/com/hellonetra/ocr/MainActivity.kt
package com.hellonetra.ocr

import android.Manifest
import android.app.Activity
import android.content.Context
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.ImageFormat
import android.graphics.Matrix
import android.graphics.RectF
import android.graphics.SurfaceTexture
import android.hardware.camera2.CameraCaptureSession
import android.hardware.camera2.CameraCharacteristics
import android.hardware.camera2.CameraDevice
import android.hardware.camera2.CameraManager
import android.hardware.camera2.CaptureRequest
import android.media.ExifInterface
import android.media.ImageReader
import android.os.Bundle
import android.os.Handler
import android.os.HandlerThread
import android.util.Size
import android.util.SparseIntArray
import android.view.Surface
import android.view.TextureView
import android.widget.Button
import android.widget.TextView
import com.paddle.ocr.EngineConfig
import com.paddle.ocr.PaddleOCR
import com.paddle.ocr.PaddleOCRConfig
import com.paddle.ocr.util.OpenCVUtils
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import java.io.ByteArrayInputStream
import kotlin.math.abs
import kotlin.math.max

class MainActivity : Activity() {

    companion object {
        private const val CAMERA_PERMISSION_REQUEST = 1001

        private val ORIENTATIONS = SparseIntArray().apply {
            append(Surface.ROTATION_0, 0)
            append(Surface.ROTATION_90, 90)
            append(Surface.ROTATION_180, 180)
            append(Surface.ROTATION_270, 270)
        }
    }

    private lateinit var textureView: TextureView
    private lateinit var scanButton: Button
    private lateinit var statusText: TextView
    private lateinit var resultText: TextView

    private val appScope = CoroutineScope(
        SupervisorJob() + Dispatchers.Main.immediate
    )

    private var ocr: PaddleOCR? = null

    private var cameraDevice: CameraDevice? = null
    private var cameraSession: CameraCaptureSession? = null
    private var cameraId: String? = null
    private var imageReader: ImageReader? = null
    private var cameraThread: HandlerThread? = null
    private var cameraHandler: Handler? = null
    private var previewSurface: Surface? = null
    private var cameraCharacteristics: CameraCharacteristics? = null
    private var previewSize: Size? = null
    private var captureSize: Size? = null

    private var scanInProgress = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        textureView = findViewById(R.id.textureView)
        scanButton = findViewById(R.id.scanButton)
        statusText = findViewById(R.id.statusText)
        resultText = findViewById(R.id.resultText)

        statusText.text = "Loading OCR models..."
        scanButton.isEnabled = false

        textureView.surfaceTextureListener = surfaceTextureListener

        scanButton.setOnClickListener {
            captureText()
        }

        loadOcr()
    }

    private fun loadOcr() {
        appScope.launch {
            try {
                statusText.text = "Initializing OpenCV..."

                val opencvReady = OpenCVUtils.init(this@MainActivity)

                if (!opencvReady) {
                    statusText.text = "OpenCV initialization failed"
                    resultText.text = "Could not load OpenCV native library."
                    return@launch
                }

                statusText.text = "Loading OCR models..."

                val engine = PaddleOCR.create(
                    context = this@MainActivity,
                    config = PaddleOCRConfig(
                        detImgMode = "BGR",
                        detLimitSideLen = 64,
                        detLimitType = "min",
                        detMaxSideLimit = 4000,
                        detThresh = 0.3f,
                        detBoxThresh = 0.6f,
                        detUnclipRatio = 1.5f,
                        detMaxCandidates = 3000,
                        detUseDilation = false,
                        detScoreMode = "fast",
                        detBoxType = "quad",
                        recScoreThresh = 0.0f,
                        recBatchSize = 1,
                    ),
                    engineConfig = EngineConfig(
                        numThreads = 4,
                    ),
                    detModelAssetPath =
                        "det/ch_PP-OCRv5_det_mobile.onnx",
                    recModelAssetPath =
                        "rec/en_PP-OCRv5_rec_mobile.onnx",
                    recConfigAssetPath =
                        "rec/ppocrv5_en_rec.yml",
                )

                ocr = engine

                statusText.text = "OCR ready"

                if (checkSelfPermission(Manifest.permission.CAMERA)
                    != PackageManager.PERMISSION_GRANTED
                ) {
                    requestPermissions(
                        arrayOf(Manifest.permission.CAMERA),
                        CAMERA_PERMISSION_REQUEST
                    )
                } else {
                    if (textureView.isAvailable) {
                        openCamera(textureView.width, textureView.height)
                    }
                }

                scanButton.isEnabled = true

            } catch (t: Throwable) {
                statusText.text =
                    "OCR load failed: ${t.message ?: t.javaClass.simpleName}"
                resultText.text = ""
            }
        }
    }

    private val surfaceTextureListener = object : TextureView.SurfaceTextureListener {

        override fun onSurfaceTextureAvailable(
            surface: SurfaceTexture,
            width: Int,
            height: Int,
        ) {
            if (checkSelfPermission(Manifest.permission.CAMERA)
                == PackageManager.PERMISSION_GRANTED
            ) {
                openCamera(width, height)
            }
        }

        override fun onSurfaceTextureSizeChanged(
            surface: SurfaceTexture,
            width: Int,
            height: Int,
        ) {
            configureTextureTransform(width, height)
        }

        override fun onSurfaceTextureDestroyed(surface: SurfaceTexture): Boolean {
            return true
        }

        override fun onSurfaceTextureUpdated(surface: SurfaceTexture) {
        }
    }

    private fun startCameraThread() {
        if (cameraThread != null) return

        cameraThread = HandlerThread("HelloNetraCamera").also {
            it.start()
            cameraHandler = Handler(it.looper)
        }
    }

    private fun openCamera(viewWidth: Int, viewHeight: Int) {
        try {
            startCameraThread()

            val manager =
                getSystemService(Context.CAMERA_SERVICE) as CameraManager

            val selectedId = manager.cameraIdList.firstOrNull { id ->
                val chars = manager.getCameraCharacteristics(id)
                chars.get(CameraCharacteristics.LENS_FACING) ==
                    CameraCharacteristics.LENS_FACING_BACK
            } ?: manager.cameraIdList.firstOrNull()

            if (selectedId == null) {
                statusText.text = "No camera found"
                return
            }

            cameraId = selectedId

            val chars = manager.getCameraCharacteristics(selectedId)
            cameraCharacteristics = chars

            val map = chars.get(
                CameraCharacteristics.SCALER_STREAM_CONFIGURATION_MAP
            )

            if (map == null) {
                statusText.text = "Camera configuration unavailable"
                return
            }

            val previewSizes =
                map.getOutputSizes(SurfaceTexture::class.java)

            val jpegSizes =
                map.getOutputSizes(ImageFormat.JPEG)

            previewSize = chooseSize(
                previewSizes,
                viewWidth,
                viewHeight
            )

            captureSize = chooseSize(
                jpegSizes,
                1280,
                720
            )

            val chosenPreviewSize = previewSize
                ?: Size(1280, 720)

            val chosenCaptureSize = captureSize
                ?: Size(1280, 720)

            textureView.surfaceTexture?.setDefaultBufferSize(
                chosenPreviewSize.width,
                chosenPreviewSize.height
            )

            configureTextureTransform(viewWidth, viewHeight)

            imageReader?.close()

            imageReader = ImageReader.newInstance(
                chosenCaptureSize.width,
                chosenCaptureSize.height,
                ImageFormat.JPEG,
                2
            )

            imageReader?.setOnImageAvailableListener(
                imageAvailableListener,
                cameraHandler
            )

            if (checkSelfPermission(Manifest.permission.CAMERA)
                != PackageManager.PERMISSION_GRANTED
            ) {
                return
            }

            manager.openCamera(
                selectedId,
                cameraStateCallback,
                cameraHandler
            )

            statusText.text = "Camera ready"

        } catch (t: Throwable) {
            statusText.text =
                "Camera error: ${t.message ?: t.javaClass.simpleName}"
        }
    }

    private fun chooseSize(
        sizes: Array<Size>?,
        targetWidth: Int,
        targetHeight: Int,
    ): Size? {
        if (sizes.isNullOrEmpty()) return null

        val targetRatio =
            targetWidth.toDouble() / targetHeight.toDouble()

        return sizes.minByOrNull { size ->
            val ratio =
                size.width.toDouble() / size.height.toDouble()

            val ratioError = abs(ratio - targetRatio)
            val sizeError =
                abs(size.width - targetWidth) +
                    abs(size.height - targetHeight)

            ratioError * 10000.0 + sizeError
        }
    }

    private val cameraStateCallback = object : CameraDevice.StateCallback() {

        override fun onOpened(camera: CameraDevice) {
            cameraDevice = camera
            createCameraSession()
        }

        override fun onDisconnected(camera: CameraDevice) {
            camera.close()

            if (cameraDevice === camera) {
                cameraDevice = null
            }
        }

        override fun onError(camera: CameraDevice, error: Int) {
            camera.close()

            if (cameraDevice === camera) {
                cameraDevice = null
            }

            statusText.text = "Camera error code: $error"
        }
    }

    private fun createCameraSession() {
        val camera = cameraDevice ?: return
        val texture = textureView.surfaceTexture ?: return
        val reader = imageReader ?: return

        try {
            val chosenPreviewSize =
                previewSize ?: Size(1280, 720)

            texture.setDefaultBufferSize(
                chosenPreviewSize.width,
                chosenPreviewSize.height
            )

            previewSurface?.release()

            previewSurface = Surface(texture)

            val sessionSurfaces = listOf(
                previewSurface!!,
                reader.surface
            )

            camera.createCaptureSession(
                sessionSurfaces,
                object : CameraCaptureSession.StateCallback() {

                    override fun onConfigured(
                        session: CameraCaptureSession
                    ) {
                        cameraSession = session
                        startPreview()
                    }

                    override fun onConfigureFailed(
                        session: CameraCaptureSession
                    ) {
                        statusText.text =
                            "Camera session failed"
                    }
                },
                cameraHandler
            )
        } catch (t: Throwable) {
            statusText.text =
                "Session error: ${t.message ?: t.javaClass.simpleName}"
        }
    }

    private fun startPreview() {
        val camera = cameraDevice ?: return
        val session = cameraSession ?: return
        val surface = previewSurface ?: return

        try {
            val request =
                camera.createCaptureRequest(
                    CameraDevice.TEMPLATE_PREVIEW
                ).apply {
                    addTarget(surface)

                    set(
                        CaptureRequest.CONTROL_AF_MODE,
                        CaptureRequest.CONTROL_AF_MODE_CONTINUOUS_PICTURE
                    )

                    set(
                        CaptureRequest.CONTROL_AE_MODE,
                        CaptureRequest.CONTROL_AE_MODE_ON
                    )
                }

            session.setRepeatingRequest(
                request.build(),
                null,
                cameraHandler
            )

            statusText.text = "Point camera at text and press SCAN"
        } catch (t: Throwable) {
            statusText.text =
                "Preview error: ${t.message ?: t.javaClass.simpleName}"
        }
    }

    private fun captureText() {
        if (scanInProgress) return

        val camera = cameraDevice ?: run {
            statusText.text = "Camera not ready"
            return
        }

        val session = cameraSession ?: run {
            statusText.text = "Camera session not ready"
            return
        }

        val reader = imageReader ?: run {
            statusText.text = "Image reader not ready"
            return
        }

        scanInProgress = true
        scanButton.isEnabled = false
        statusText.text = "Capturing..."

        try {
            val request =
                camera.createCaptureRequest(
                    CameraDevice.TEMPLATE_STILL_CAPTURE
                ).apply {
                    addTarget(reader.surface)

                    set(
                        CaptureRequest.CONTROL_AF_MODE,
                        CaptureRequest.CONTROL_AF_MODE_CONTINUOUS_PICTURE
                    )

                    set(
                        CaptureRequest.CONTROL_AE_MODE,
                        CaptureRequest.CONTROL_AE_MODE_ON
                    )

                    cameraCharacteristics?.get(
                        CameraCharacteristics.SENSOR_ORIENTATION
                    )?.let { sensorOrientation ->
                        set(
                            CaptureRequest.JPEG_ORIENTATION,
                            getJpegOrientation(sensorOrientation)
                        )
                    }
                }

            session.capture(
                request.build(),
                object : CameraCaptureSession.CaptureCallback() {
                },
                cameraHandler
            )

        } catch (t: Throwable) {
            scanInProgress = false
            scanButton.isEnabled = true

            statusText.text =
                "Capture failed: ${t.message ?: t.javaClass.simpleName}"
        }
    }

    private val imageAvailableListener =
        ImageReader.OnImageAvailableListener { reader ->

            val image = reader.acquireLatestImage()
                ?: return@OnImageAvailableListener

            try {
                val buffer = image.planes[0].buffer
                val bytes = ByteArray(buffer.remaining())

                buffer.get(bytes)

                val bitmap = decodeJpeg(bytes)

                if (bitmap == null) {
                    runOnUiThread {
                        scanInProgress = false
                        scanButton.isEnabled = true
                        statusText.text = "Could not decode camera image"
                    }
                    return@OnImageAvailableListener
                }

                appScope.launch {
                    try {
                        statusText.text = "Running OCR..."

                        val engine = ocr

                        if (engine == null) {
                            statusText.text = "OCR engine unavailable"
                            return@launch
                        }

                        val result = engine.recognize(bitmap)

                        if (result.results.isEmpty()) {
                            resultText.text = "No text detected."
                            statusText.text = "Detected 0 line(s) • ${result.totalTimeMs} ms"
                        } else {
                            resultText.text =
                                result.results.joinToString("\n") {
                                    it.text
                                }

                            statusText.text =
                                "Detected ${result.lineCount} line(s) • ${result.totalTimeMs} ms"
                        }

                    } catch (t: Throwable) {
                        resultText.text =
                            "OCR error:\n${t.message ?: t.javaClass.simpleName}"

                        statusText.text = "OCR failed"

                    } finally {
                        if (!bitmap.isRecycled) {
                            bitmap.recycle()
                        }

                        scanInProgress = false
                        scanButton.isEnabled = true
                    }
                }

            } catch (t: Throwable) {
                runOnUiThread {
                    scanInProgress = false
                    scanButton.isEnabled = true
                    statusText.text =
                        "Image processing failed: ${t.message ?: t.javaClass.simpleName}"
                }
            } finally {
                image.close()
            }
        }

    private fun decodeJpeg(bytes: ByteArray): Bitmap? {
        val bitmap =
            BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
                ?: return null

        return try {
            val exif =
                ExifInterface(ByteArrayInputStream(bytes))

            val orientation =
                exif.getAttributeInt(
                    ExifInterface.TAG_ORIENTATION,
                    ExifInterface.ORIENTATION_NORMAL
                )

            when (orientation) {
                ExifInterface.ORIENTATION_ROTATE_90 ->
                    rotateBitmap(bitmap, 90f)

                ExifInterface.ORIENTATION_ROTATE_180 ->
                    rotateBitmap(bitmap, 180f)

                ExifInterface.ORIENTATION_ROTATE_270 ->
                    rotateBitmap(bitmap, 270f)

                ExifInterface.ORIENTATION_FLIP_HORIZONTAL ->
                    flipBitmap(bitmap, horizontal = true)

                ExifInterface.ORIENTATION_FLIP_VERTICAL ->
                    flipBitmap(bitmap, horizontal = false)

                else -> bitmap
            }
        } catch (_: Throwable) {
            bitmap
        }
    }

    private fun rotateBitmap(
        bitmap: Bitmap,
        degrees: Float
    ): Bitmap {
        val matrix = Matrix().apply {
            postRotate(degrees)
        }

        val rotated = Bitmap.createBitmap(
            bitmap,
            0,
            0,
            bitmap.width,
            bitmap.height,
            matrix,
            true
        )

        if (rotated !== bitmap) {
            bitmap.recycle()
        }

        return rotated
    }

    private fun flipBitmap(
        bitmap: Bitmap,
        horizontal: Boolean
    ): Bitmap {
        val matrix = Matrix().apply {
            postScale(
                if (horizontal) -1f else 1f,
                if (horizontal) 1f else -1f
            )
        }

        val flipped = Bitmap.createBitmap(
            bitmap,
            0,
            0,
            bitmap.width,
            bitmap.height,
            matrix,
            true
        )

        if (flipped !== bitmap) {
            bitmap.recycle()
        }

        return flipped
    }

    private fun getJpegOrientation(sensorOrientation: Int): Int {
        val rotation = if (android.os.Build.VERSION.SDK_INT >= 30) {
            display?.rotation ?: Surface.ROTATION_0
        } else {
            @Suppress("DEPRECATION")
            windowManager.defaultDisplay.rotation
        }
        val deviceOrientation = ORIENTATIONS.get(rotation)
        val isFront = cameraCharacteristics?.get(CameraCharacteristics.LENS_FACING) == CameraCharacteristics.LENS_FACING_FRONT
        return if (isFront) {
            (sensorOrientation + deviceOrientation) % 360
        } else {
            (sensorOrientation - deviceOrientation + 360) % 360
        }
    }

    private fun configureTextureTransform(
        viewWidth: Int,
        viewHeight: Int
    ) {
        val size = previewSize ?: return

        if (viewWidth == 0 || viewHeight == 0) return

        val rotation = if (android.os.Build.VERSION.SDK_INT >= 30) {
            display?.rotation ?: Surface.ROTATION_0
        } else {
            @Suppress("DEPRECATION")
            windowManager.defaultDisplay.rotation
        }

        val matrix = Matrix()

        val viewRect = RectF(
            0f,
            0f,
            viewWidth.toFloat(),
            viewHeight.toFloat()
        )

        val bufferRect = RectF(
            0f,
            0f,
            size.height.toFloat(),
            size.width.toFloat()
        )

        val centerX = viewRect.centerX()
        val centerY = viewRect.centerY()

        bufferRect.offset(
            centerX - bufferRect.centerX(),
            centerY - bufferRect.centerY()
        )

        if (
            rotation == Surface.ROTATION_90 ||
            rotation == Surface.ROTATION_270
        ) {
            matrix.setRectToRect(
                viewRect,
                bufferRect,
                Matrix.ScaleToFit.FILL
            )

            val scale = max(
                viewHeight.toFloat() / size.width.toFloat(),
                viewWidth.toFloat() / size.height.toFloat()
            )

            matrix.postScale(
                scale,
                scale,
                centerX,
                centerY
            )

            matrix.postRotate(
                90f * (rotation - 2),
                centerX,
                centerY
            )

        } else if (rotation == Surface.ROTATION_180) {
            matrix.postRotate(
                180f,
                centerX,
                centerY
            )
        }

        textureView.setTransform(matrix)
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(
            requestCode,
            permissions,
            grantResults
        )

        if (requestCode == CAMERA_PERMISSION_REQUEST) {
            if (
                grantResults.isNotEmpty() &&
                grantResults[0] == PackageManager.PERMISSION_GRANTED
            ) {
                statusText.text = "Camera permission granted"

                if (textureView.isAvailable) {
                    openCamera(
                        textureView.width,
                        textureView.height
                    )
                }
            } else {
                statusText.text =
                    "Camera permission is required"
            }
        }
    }

    override fun onDestroy() {
        super.onDestroy()

        cameraSession?.close()
        cameraSession = null

        cameraDevice?.close()
        cameraDevice = null

        imageReader?.close()
        imageReader = null

        previewSurface?.release()
        previewSurface = null

        cameraThread?.quitSafely()
        cameraThread = null
        cameraHandler = null

        appScope.launch {
            try {
                ocr?.release()
            } catch (_: Throwable) {
            }
            ocr = null
        }

        appScope.cancel()
    }
}